File: test-decrypt-priya.js
#!/usr/bin/env node

/**
 * Encryption Key Test & Decryption
 * Tests the provided encryption key for Priya Mishra's backup
 * Node.js version
 */

import fs from 'fs';
import path from 'path';
import crypto from 'crypto';
import { createClient } from '@supabase/supabase-js';
import 'dotenv/config';

class DecryptionTester {
  constructor(encryptionKey, phoneNumber, name) {
    this.encryptionKey = encryptionKey;
    this.phoneNumber = phoneNumber;
    this.name = name;
    
    // Backup directory
    this.backupDir = '/home/katyayani/Downloads/AndroidData-20260310T103958Z-3-001/AndroidData';
    
    // Initialize Supabase for updating if key works
    const supabaseUrl = process.env.SUPABASE_URL;
    const supabaseKey = process.env.SUPABASE_KEY || 
                       process.env.SUPABASE_SERVICE_KEY || 
                       process.env.SUPABASE_ANON_KEY;

    if (supabaseUrl && supabaseKey) {
      this.supabase = createClient(supabaseUrl, supabaseKey);
      console.log(`✅ Supabase connected\n`);
    }
  }

  /**
   * Extract IV from Crypt15 backup protobuf header
   */
  extractIVFromProtobuf(encryptedData) {
    try {
      const protobufSize = encryptedData[0];
      let offset = 1;

      // Check for feature flag (0x01)
      if (encryptedData[offset] === 0x01) {
        offset += 1;
      }

      // Extract protobuf data
      const protobufData = encryptedData.slice(offset, offset + protobufSize);

      // Look for the IV in the protobuf (field with 0x0a prefix and 16-byte length)
      for (let i = 0; i < protobufData.length - 17; i++) {
        if (protobufData[i] === 0x0a && protobufData[i + 1] === 16) {
          const candidate = protobufData.slice(i + 2, i + 18);
          if (candidate.length === 16) {
            return candidate;
          }
        }
      }

      // Fallback: find any 16-byte sequence with good entropy
      for (let i = 0; i < protobufData.length - 15; i++) {
        const candidate = protobufData.slice(i, i + 16);
        const isZero = candidate.every(b => b === 0);
        const allSame = candidate.slice(1).every(b => b === candidate[0]);
        
        if (!isZero && !allSame && candidate.length === 16) {
          return candidate;
        }
      }

      return null;
    } catch (e) {
      console.log(`Warning: IV extraction error: ${e.message}`);
      return null;
    }
  }

  /**
   * Test decryption with the provided key
   */
  async testDecryption(filePath) {
    const fileName = path.basename(filePath);
    console.log(`\n🔓 Testing decryption of: ${fileName}`);

    try {
      // Read encrypted file
      const encryptedData = fs.readFileSync(filePath);
      const fileSize = encryptedData.length;
      console.log(`   File size: ${(fileSize / 1024).toFixed(2)} KB`);

      // Validate key format
      let keyBytes;
      try {
        keyBytes = Buffer.from(this.encryptionKey, 'hex');
        if (keyBytes.length !== 32) {
          console.log(`❌ Invalid key length: ${keyBytes.length} bytes (expected 32)`);
          return false;
        }
      } catch (e) {
        console.log(`❌ Invalid key format: ${e.message}`);
        return false;
      }

      // Extract IV from backup
      const iv = this.extractIVFromProtobuf(encryptedData);
      if (!iv || iv.length !== 16) {
        console.log(`❌ Could not extract IV from backup header`);
        return false;
      }

      console.log(`   IV: ${iv.toString('hex')}`);

      // Extract encryption components
      const checksum = encryptedData.slice(-16);
      const authTag = encryptedData.slice(-32, -16);
      const encryptedPayload = encryptedData.slice(0, -32);

      console.log(`   Payload size: ${(encryptedPayload.length / 1024).toFixed(2)} KB`);

      // Attempt AES-GCM decryption
      const decipher = crypto.createDecipheriv('aes-256-gcm', keyBytes, iv.slice(0, 12));
      
      // Set auth tag BEFORE calling final()
      decipher.setAuthTag(authTag);
      
      let decrypted;
      try {
        decrypted = Buffer.concat([
          decipher.update(encryptedPayload),
          decipher.final()
        ]);
        
        console.log(`\n✅ DECRYPTION SUCCESSFUL!`);
        console.log(`   ✓ Authentication tag verified`);
        console.log(`   ✓ Decrypted size: ${(decrypted.length / 1024).toFixed(2)} KB`);
        
        // Check for zlib compression
        if (decrypted[0] === 0x78) {
          const zlib = await import('zlib');
          try {
            decrypted = zlib.inflateSync(decrypted);
            console.log(`   ✓ Decompressed: ${(decrypted.length / 1024).toFixed(2)} KB`);
          } catch (e) {
            console.log(`   ⚠️  Decompression failed: ${e.message}`);
          }
        }

        // Check if SQLite
        if (decrypted.length > 16 && decrypted.slice(0, 13).toString() === 'SQLite format') {
          console.log(`   ✅ Valid SQLite database file!`);
        } else {
          console.log(`   ⚠️  First 16 bytes: ${decrypted.slice(0, 16).toString('hex')}`);
        }

        // Save decrypted file
        const outputFileName = `decrypted_${path.basename(filePath, '.crypt15')}.db`;
        const outputPath = path.join(process.cwd(), outputFileName);
        fs.writeFileSync(outputPath, decrypted);
        console.log(`   💾 Saved: ${outputFileName}`);

        return {
          success: true,
          fileName: fileName,
          decryptedSize: decrypted.length,
          outputPath: outputPath
        };
        
      } catch (e) {
        console.log(`\n❌ DECRYPTION FAILED!`);
        console.log(`   ✗ Authentication tag mismatch: ${e.message}`);
        console.log(`\n   This key does NOT match the encryption used for this backup.`);
        return false;
      }

    } catch (e) {
      console.log(`❌ Error: ${e.message}`);
      return false;
    }
  }

  /**
   * Main test execution
   */
  async run() {
    console.log(`
╔═══════════════════════════════════════════════════════════════╗
║              Encryption Key Decryption Tester v1.0             ║
║         Testing provided encryption key for backup files       ║
╚═══════════════════════════════════════════════════════════════╝`);

    console.log(`\n📱 Phone: ${this.phoneNumber}`);
    console.log(`👤 Name: ${this.name}`);
    console.log(`🔑 Key: ${this.encryptionKey.substring(0, 32)}...${this.encryptionKey.substring(-8)}`);

    // Check backup directory
    if (!fs.existsSync(this.backupDir)) {
      console.log(`\n❌ Backup directory not found: ${this.backupDir}`);
      return;
    }

    // Find backup files
    const backupFiles = fs.readdirSync(this.backupDir)
      .filter(f => f.endsWith('.crypt15'))
      .map(f => path.join(this.backupDir, f));

    if (backupFiles.length === 0) {
      console.log(`\n❌ No .crypt15 backup files found`);
      return;
    }

    console.log(`\n📁 Found ${backupFiles.length} backup files:`);
    backupFiles.forEach((f, i) => {
      const stats = fs.statSync(f);
      console.log(`   ${i + 1}. ${path.basename(f)} - ${(stats.size / 1024).toFixed(2)} KB`);
    });

    // Test decryption of all files
    console.log(`\n${'='.repeat(70)}`);
    console.log(`🔑 Testing Encryption Key`);
    console.log('='.repeat(70));

    const results = [];
    for (const file of backupFiles) {
      const result = await this.testDecryption(file);
      if (result && result.success) {
        results.push(result);
      }
    }

    // Print summary
    this.printSummary(results);
  }

  /**
   * Print summary of decryption results
   */
  printSummary(results) {
    console.log(`\n\n╔═══════════════════════════════════════════════════════════════╗`);
    console.log(`║                    DECRYPTION SUMMARY                         ║`);
    console.log(`╠═══════════════════════════════════════════════════════════════╣`);

    if (results.length === 0) {
      console.log(`║  ❌ All decryption attempts FAILED                            ║`);
      console.log(`║     The provided key does NOT match this backup               ║`);
      console.log(`║                                                               ║`);
      console.log(`║  Next steps:                                                  ║`);
      console.log(`║  1. Double-check the encryption key is correct                ║`);
      console.log(`║  2. Extract key from WhatsApp Settings → Chats → Backup       ║`);
      console.log(`║  3. Or extract from phone via ADB from WhatsApp database      ║`);
    } else {
      console.log(`║  ✅ Successfully decrypted ${results.length} file(s)!                    ║`);
      console.log(`╠═══════════════════════════════════════════════════════════════╣`);
      
      for (const result of results) {
        console.log(`║                                                               ║`);
        console.log(`║  File: ${result.fileName.padEnd(55)} ║`);
        console.log(`║  Size: ${(result.decryptedSize / 1024).toFixed(2)} KB${' '.repeat(58 - 10)} ║`);
        console.log(`║  Location: ${result.outputPath.padEnd(50)} ║`);
      }

      console.log(`║                                                               ║`);
      console.log(`║  ✅ KEY IS VALID!                                             ║`);
      console.log(`║                                                               ║`);
      console.log(`║  You can now:                                                 ║`);
      console.log(`║  1. Use the decrypted database files for analysis             ║`);
      console.log(`║  2. Update Supabase with this confirmed key:                  ║`);
      console.log(`║     UPDATE phones SET encryption_key = '${this.encryptionKey}' ║`);
      console.log(`║     WHERE phone_number = '${this.phoneNumber}';                ║`);
      console.log(`║  3. Parse the SQLite databases for messages/chats             ║`);
    }

    console.log(`╚═══════════════════════════════════════════════════════════════╝`);
  }
}

// Main execution
async function main() {
  // Encryption key provided by user
  const encryptionKey = 'aca75852758afeaad9c5b1ed81b889daf16b77600c028e33aebb153214ce839a';
  const phoneNumber = 'Priya Mishra';
  const name = 'Priya Mishra';

  try {
    const tester = new DecryptionTester(encryptionKey, phoneNumber, name);
    await tester.run();
  } catch (e) {
    console.log(`\n❌ Fatal error: ${e.message}`);
    console.error(e);
    process.exit(1);
  }
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch(e => {
    console.error('Fatal error:', e);
    process.exit(1);
  });
}

export default DecryptionTester;

File: test-decrypt-priya.py
#!/usr/bin/env python3

"""
Test decryption with provided encryption key for Priya Mishra's backup
Python version using wa-crypt-tools for comparison
"""

import os
import sys
import io
import zlib
from pathlib import Path
from dotenv import load_dotenv

# Load environment
load_dotenv()

# Import wa-crypt-tools
try:
    from wa_crypt_tools.lib.key.key15 import Key15
    from wa_crypt_tools.lib.db.db15 import Database15
    from Cryptodome.Cipher import AES
except ImportError:
    print("❌ wa-crypt-tools not installed!")
    sys.exit(1)


class DecryptionTesterPython:
    """Test decryption with provided encryption key"""
    
    def __init__(self, encryption_key_hex, phone_name):
        self.encryption_key_hex = encryption_key_hex
        self.phone_name = phone_name
        self.backup_dir = Path("/home/katyayani/Downloads/AndroidData-20260310T103958Z-3-001/AndroidData")
    
    def test_file(self, file_path):
        """Test decryption of a single file"""
        
        print(f"\n🔓 Testing: {file_path.name}")
        
        try:
            # Read encrypted file
            with open(file_path, 'rb') as f:
                encrypted_data = f.read()
            
            print(f"   File size: {len(encrypted_data) / 1024:.2f} KB")
            
            # Initialize key
            key_bytes = bytes.fromhex(self.encryption_key_hex)
            key15 = Key15(keyarray=key_bytes)
            
            # Parse header
            encrypted_stream = io.BufferedReader(io.BytesIO(encrypted_data))
            db = Database15(key=key15, encrypted=encrypted_stream)
            
            # Get IV
            iv = db.header.c15_iv.IV
            print(f"   IV: {iv.hex()}")
            
            # Extract components
            checksum = encrypted_data[-16:]
            auth_tag = encrypted_data[-32:-16]
            encrypted_payload = encrypted_data[:-32]
            
            # Try decryption
            encryption_key = key15.get()
            cipher = AES.new(encryption_key, AES.MODE_GCM, iv[:12])
            
            decrypted = cipher.decrypt(encrypted_payload)
            
            try:
                cipher.verify(auth_tag)
                
                # Check for zlib
                if decrypted[0:1] == b'\x78':
                    decrypted = zlib.decompress(decrypted)
                    print(f"   ✓ Decompressed to {len(decrypted) / 1024:.2f} KB")
                
                # Check if SQLite
                if decrypted[:13] == b'SQLite format':
                    print(f"   ✅ DECRYPTION SUCCESS - Valid SQLite!")
                    return True
                else:
                    print(f"   ✅ DECRYPTION SUCCESS - {len(decrypted)} bytes")
                    print(f"   First bytes: {decrypted[:16].hex()}")
                    return True
                    
            except ValueError as e:
                print(f"   ❌ Auth tag failed: {e}")
                return False
                
        except Exception as e:
            print(f"   ❌ Error: {e}")
            import traceback
            traceback.print_exc()
            return False
    
    def run(self):
        """Main execution"""
        
        print(f"""
╔═══════════════════════════════════════════════════════════════╗
║     Encryption Key Decryption Tester (Python + wa-crypt-tools) ║
║           Testing provided encryption key                      ║
╚═══════════════════════════════════════════════════════════════╝""")
        
        print(f"\n👤 Name: {self.phone_name}")
        print(f"🔑 Key: {self.encryption_key_hex[:32]}...{self.encryption_key_hex[-8:]}")
        
        # Check backup dir
        if not self.backup_dir.exists():
            print(f"❌ Backup directory not found: {self.backup_dir}")
            return
        
        # Find files
        backup_files = sorted(self.backup_dir.glob("*.crypt15"))
        
        if not backup_files:
            print(f"❌ No .crypt15 files found")
            return
        
        print(f"\n📁 Found {len(backup_files)} backup files:")
        for i, f in enumerate(backup_files, 1):
            size = f.stat().st_size / 1024
            print(f"   {i}. {f.name} - {size:.2f} KB")
        
        # Test decryption
        print(f"\n{'='*70}")
        print(f"🔑 Testing Encryption Key with wa-crypt-tools")
        print('='*70)
        
        results = []
        for file_path in backup_files:
            success = self.test_file(file_path)
            if success:
                results.append(file_path.name)
        
        # Summary
        print(f"\n\n╔═══════════════════════════════════════════════════════════════╗")
        print(f"║                    TEST SUMMARY                               ║")
        print(f"╠═══════════════════════════════════════════════════════════════╣")
        
        if results:
            print(f"║  ✅ Successfully decrypted {len(results)} file(s)!                    ║")
            print(f"║                                                               ║")
            for fname in results:
                print(f"║     ✓ {fname.ljust(58)} ║")
            print(f"║                                                               ║")
            print(f"║  🎉 KEY IS VALID!                                             ║")
        else:
            print(f"║  ❌ All decryption attempts FAILED                            ║")
            print(f"║                                                               ║")
            print(f"║  The provided key does NOT match this backup                 ║")
            print(f"║                                                               ║")
            print(f"║  Possible causes:                                             ║")
            print(f"║  1. Key is incorrect or outdated                             ║")
            print(f"║  2. Key was changed after backup creation                    ║")
            print(f"║  3. Backup is from a different encryption root key           ║")
            print(f"║                                                               ║")
            print(f"║  To get the correct key:                                     ║")
            print(f"║  1. On the phone: Settings → Chats → Chat backup             ║")
            print(f"║  2. Look for 'Encryption' section - 64 hex character key      ║")
            print(f"║  3. Or extract via ADB from WhatsApp database files          ║")
        
        print(f"╚═══════════════════════════════════════════════════════════════╝")


def main():
    encryption_key = "aca75852758afeaad9c5b1ed81b889daf16b77600c028e33aebb153214ce839a"
    phone_name = "Priya Mishra"
    
    try:
        tester = DecryptionTesterPython(encryption_key, phone_name)
        tester.run()
    except Exception as e:
        print(f"\n❌ Fatal error: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)


if __name__ == "__main__":
    main()
